lib/storage/authgateway_test.go

```diff
@@ -0,0 +1,244 @@
+			}),
+			gw2: NewAuthGateway(AuthGatewaySpecV2{
+				SSHPublicAddr: &[]string{"ssh.example.com"},
+			}),
+			result: true,
+		},
+		{
+			gw1: NewAuthGateway(AuthGatewaySpecV2{
+				PublicAddr: &[]string{"example.com"},
+			}),
+			gw2: NewAuthGateway(AuthGatewaySpecV2{
+				SSHPublicAddr:        &[]string{"example.com"},
+				KubernetesPublicAddr: &[]string{"example.com"},
+				WebPublicAddr:        &[]string{"example.com"},
+			}),
+			result: false,
+		},
+		{
+			gw1: NewAuthGateway(AuthGatewaySpecV2{
+				PublicAddr: &[]string{"example.com"},
+			}),
+			gw2: NewAuthGateway(AuthGatewaySpecV2{
+				SSHPublicAddr:        &[]string{"example.com"},
+				KubernetesPublicAddr: &[]string{"example.com"},
+			}),
+			result: true,
+		},
+		{
+			gw1: NewAuthGateway(AuthGatewaySpecV2{
+				KubernetesPublicAddr: &[]string{"k8s.example.com:3036"},
+			}),
+			gw2: NewAuthGateway(AuthGatewaySpecV2{
+				KubernetesPublicAddr: &[]string{"k8s.example.com:3027"},
+			}),
+			result: false,
+		},
+	}
+	for _, tc := range testCases {
+		c.Assert(tc.gw1.PrincipalsChanged(tc.gw2), check.Equals, tc.result,
+			check.Commentf("Test case %s/%s/%v failed.", tc.gw1, tc.gw2, tc.result))
+	}
+}
+
+func (s *AuthGatewaySuite) TestSettingsChanged(c *check.C) {
+	testCases := []struct {
+		gw1, gw2 AuthGateway
+		result   bool
+	}{
+		{
+			gw1: NewAuthGateway(AuthGatewaySpecV2{
+				ConnectionLimits: &ConnectionLimits{
+					MaxConnections: int64p(1000),
+					MaxUsers:       intp(10),
+				},
+			}),
+			gw2: NewAuthGateway(AuthGatewaySpecV2{
+				ConnectionLimits: &ConnectionLimits{
+					MaxConnections: int64p(1500),
+					MaxUsers:       intp(10),
+				},
+			}),
+			result: true,
+		},
+		{
+			gw1: NewAuthGateway(AuthGatewaySpecV2{
+				SSHPublicAddr: &[]string{"example.com"},
+			}),
+			gw2: NewAuthGateway(AuthGatewaySpecV2{
+				SSHPublicAddr: &[]string{"ssh.example.com"},
+			}),
+			result: false,
+		},
+	}
+	for _, tc := range testCases {
+		c.Assert(tc.gw1.SettingsChanged(tc.gw2), check.Equals, tc.result,
+			check.Commentf("Test case %s/%s/%v failed.", tc.gw1, tc.gw2, tc.result))
+	}
+}
+
+func (s *AuthGatewaySuite) TestApplyTo(c *check.C) {
+	gw1 := NewAuthGateway(AuthGatewaySpecV2{
+		ConnectionLimits: &ConnectionLimits{
+			MaxConnections: int64p(1000),
+			MaxUsers:       intp(10),
+		},
+		SSHPublicAddr: &[]string{"ssh.example.com"},
+	})
+	gw2 := NewAuthGateway(AuthGatewaySpecV2{
+		ConnectionLimits: &ConnectionLimits{
+			MaxUsers: intp(5),
+		},
+		PublicAddr: &[]string{"example.com"},
+	})
+	gw2.ApplyTo(gw1)
+	c.Assert(gw1, compare.DeepEquals, NewAuthGateway(AuthGatewaySpecV2{
+		ConnectionLimits: &ConnectionLimits{
+			MaxConnections: int64p(1000),
+			MaxUsers:       intp(5),
+		},
+		SSHPublicAddr:        &[]string{"example.com"},
+		KubernetesPublicAddr: &[]string{"example.com"},
+		WebPublicAddr:        &[]string{"example.com"},
+	}))
+}
+
+func (s *AuthGatewaySuite) TestApplyToTeleportConfig(c *check.C) {
```

Comment:
Frankly, I would've replaced all these tests with a single integration test that executes the event loops and validates that the configuration is properly received/updated.

Reply:
I think unit tests have their own merit. I've added one more higher-level test to test configuration update though (in process_test.go).